mm2src/coins/eth/eth_swap_v2.rs

```diff
@@ -137,14 +543,404 @@ impl EthCoin {
         ])?;
         Ok(data)
     }
+
+    /// Prepares data for EtomicSwapTakerV2 contract `refundTakerPaymentTimelock` method
+    /// function refundMakerPaymentTimelock(
+    ///         bytes32 id,
+    ///         uint256 amount,
+    ///         address taker,
+    ///         bytes32 takerSecretHash,
+    ///         bytes32 makerSecretHash,
+    ///         address tokenAddress
+    async fn prepare_taker_refund_payment_timelock_data(
+        &self,
+        args: TakerRefundArgs,
+    ) -> Result<Vec<u8>, PrepareTxDataError> {
+        let function = TAKER_SWAP_V2.function("refundTakerPaymentTimelock")?;
+        let id = self.etomic_swap_id_v2(args.payment_time_lock, &args.maker_secret_hash);
+        let data = function.encode_input(&[
+            Token::FixedBytes(id),
+            Token::Uint(args.payment_amount),
+            Token::Uint(args.dex_fee),
+            Token::Address(args.maker_address),
+            Token::FixedBytes(args.taker_secret_hash.to_vec()),
+            Token::FixedBytes(args.maker_secret_hash.to_vec()),
+            Token::Address(args.token_address),
+        ])?;
+        Ok(data)
+    }
+
+    /// Prepares data for EtomicSwapTakerV2 contract `refundTakerPaymentSecret` method
+    /// function refundTakerPaymentSecret(
+    ///         bytes32 id,
+    ///         uint256 amount,
+    ///         uint256 dexFee,
+    ///         address maker,
+    ///         bytes32 takerSecret,
+    ///         bytes32 makerSecretHash,
+    ///         address tokenAddress
+    async fn prepare_taker_refund_payment_secret_data(
+        &self,
+        args: &TakerRefundArgs,
+    ) -> Result<Vec<u8>, PrepareTxDataError> {
+        let function = TAKER_SWAP_V2.function("refundTakerPaymentSecret")?;
+        let id = self.etomic_swap_id_v2(args.payment_time_lock, &args.maker_secret_hash);
+        let data = function.encode_input(&[
+            Token::FixedBytes(id),
+            Token::Uint(args.payment_amount),
+            Token::Uint(args.dex_fee),
+            Token::Address(args.maker_address),
+            Token::FixedBytes(args.taker_secret.to_vec()),
+            Token::FixedBytes(args.maker_secret_hash.to_vec()),
+            Token::Address(args.token_address),
+        ])?;
+        Ok(data)
+    }
+
+    /// This function constructs the encoded transaction input data required to approve the taker payment.
+    /// The `decoded` parameter should contain the transaction input data from the `ethTakerPayment` or `erc20TakerPayment` function of the EtomicSwapTakerV2 contract.
+    async fn prepare_taker_payment_approve_data(
+        &self,
+        args: &GenTakerFundingSpendArgs<'_, Self>,
+        decoded: Vec<Token>,
```

Comment:
Shouldn't we do sanity check for `decoded` length (before we access its elements)?

Reply:
ethabi crate already does it in source code using `offset`
```rust
/// Decodes ABI compliant vector of bytes into vector of tokens described by types param.
pub fn decode(types: &[ParamType], data: &[u8]) -> Result<Vec<Token>, Error> {
	let is_empty_bytes_valid_encoding = types.iter().all(|t| t.is_empty_bytes_valid_encoding());
	if !is_empty_bytes_valid_encoding && data.is_empty() {
		return Err(Error::InvalidName(
			"please ensure the contract and method you're calling exist! \
			 failed to decode empty bytes. if you're using jsonrpc this is \
			 likely due to jsonrpc returning `0x` in case contract or method \
			 don't exist"
				.into(),
		));
	}

	let mut tokens = vec![];
	let mut offset = 0;

	for param in types {
		let res = decode_param(param, data, offset)?;
		offset = res.new_offset;
		tokens.push(res.token);
	}

	Ok(tokens)
}
```
you just need to be sure that you dont use decoded from unrelated smart contract function.
or you would like to do additional check?

Reply:
@dimxy added decoded length check https://github.com/KomodoPlatform/komodo-defi-framework/pull/2169/commits/2381b7ded8c83d19fe6462e0a09ab923e1891d2d